Add a sample program that outputs the parse info for each input file

thePHTest opened the pull request:
Open to adjusting the examples and renaming things as needed. Maybe this would make more sense under a tools folder instead of as a sample?

ryanfleury reviewed (changes requested):
Just added `MD_StringFromNodeKind`, `MD_StringListFromNodeFlags`, and `MD_ArrayCount`. Could we use those in the sample program to avoid having to maintain two lists for node kinds and flags?

thePHTest commented:
Rebased to get rid of those extraneous commits showing up here from the master branch and synced the code with your updates. I also added MD_JoinStringWithSeparator(...). Hmm, maybe I should add a test for that actually.
Also, one thing I noted in the parse output is that @tags with parameters don't have the parameters show up in the output. Is that expected?

ryanfleury commented:
Yeah, ideally it'd show tag arguments as well. Those are just children of tag nodes, so you can just do:

```
for(MD_EachNode(arg, tag->first_child))
{
  // arg is a tag argument of tag
}
```

Could we also use `MD_NodeFlags` here instead of `MD_u32`?: https://github.com/ryanfleury/data_desk/blob/86203a41f845f7d777a6f66236b14ec952cd64f4/samples/output_parse/output_parse.c#L20

The code at https://github.com/ryanfleury/data_desk/blob/86203a41f845f7d777a6f66236b14ec952cd64f4/samples/output_parse/output_parse.c#L60 could be cleaned up a bit also:

```
for(MD_EachNode(root, first))
{
  MD_String8 code_filename = MD_TrimExtension(MD_TrimFolder(root->filename));
  MD_String8 info_filename = MD_PushStringF("parsed_%.*s.txt", MD_StringExpand(code_filename));
  printf("Parse Input -> Output: %.*s -> %.*s\n", MD_StringExpand(code_filename), MD_StringExpand(info_filename));
        
  FILE* file = fopen((char *)info_filename.str, "w");
  for(MD_EachNode(node, root->first_child))
  {
    PrintNode(node, file, 0);
    fprintf(file, "\n");
  }
  fclose(file);
}
```

thePHTest commented:
Ah, good catch. Much cleaner string building that way. Thanks for the feedback. I added printing the tag parameters in the output now too.

thePHTest commented on samples/output_parse/output_parse.c:
```diff
@@ -2,16 +2,22 @@
+    for (int i = 0; i < INDENT_SPACES; i++) {
+        one_indent[i] = ' ';
+    }
+    one_indent[INDENT_SPACES] = '\0';
```
I suppose one_indent could be definted by the preprocessor somehow or as a static variable elsewhere. Doesn't need to be defined and every time this function recurses.

ryanfleury replied:
Yeah, or you can probably have a helper function like so:

```
static void Print(FILE *file, int indent_spaces, char *fmt, ...)
{
  for(int i = 0; i < indent_spaces; i += 1)
  {
    fprintf(file, " ");
  }
  va_list args;
  va_start(args, fmt);
  vfprintf(file, fmt, args);
  va_end(args);
}
```

thePHTest replied:
yeah, that's much cleaner. Clearly I haven't done much string manipulation in C. Thanks!
